File: cc/skottie.cc
#include "shade/cc/skottie.h"
#include "include/core/SkSize.h"
#include "modules/skottie/include/Skottie.h"
#include "shade/src/skottie.rs.h"

Animation::Animation(sk_sp<skottie::Animation> animation)
    : animation_(std::move(animation)) {}

double Animation::duration() const {
  return animation_->duration();
}

double Animation::fps() const {
  return animation_->fps();
}

void Animation::size(float& width, float& height) const {
  SkSize size = animation_->size();
  width = size.fWidth;
  height = size.fHeight;
}

void Animation::render(SkCanvas& canvas) const {
  animation_->render(&canvas);
}

void Animation::seek_frame(double t) {
  animation_->seekFrame(t);
}

void Animation::seek_frame_time(double t) {
  animation_->seekFrameTime(t);
}

std::unique_ptr<Animation> new_animation(rust::Slice<const std::uint8_t> data) {
  if (data.size() == 0) {
    return nullptr;
  }

  sk_sp<skottie::Animation> animation = skottie::Animation::Builder().make(
      reinterpret_cast<const char*>(data.data()), data.size());
  if (!animation) {
    return nullptr;
  }

  return std::make_unique<Animation>(std::move(animation));
}

File: src/skottie.rs
#[cxx::bridge]
mod ffi {
    unsafe extern "C++" {
        include!("shade/cc/skottie.h");

        type Animation;

        fn new_animation(data: &[u8]) -> UniquePtr<Animation>;

        fn duration(self: &Animation) -> f64;

        fn fps(self: &Animation) -> f64;

        type SkSize = crate::core::SkSize;

        fn size(self: &Animation) -> SkSize;

        type SkCanvas = crate::core::SkCanvas;

        fn render(self: &Animation, canvas: Pin<&mut SkCanvas>);

        fn seek_frame(self: Pin<&mut Animation>, t: f64);

        fn seek_frame_time(self: Pin<&mut Animation>, t: f64);
    }
}

pub use self::ffi::*;

impl Animation {
    #[inline]
    pub fn new<T: AsRef<[u8]>>(data: T) -> cxx::UniquePtr<Self> {
        new_animation(data.as_ref())
    }
}
